include/grpc/event_engine/memory_allocator.h

```diff
@@ -0,0 +1,242 @@
+// limitations under the License.
+#ifndef GRPC_EVENT_ENGINE_MEMORY_ALLOCATOR_H
+#define GRPC_EVENT_ENGINE_MEMORY_ALLOCATOR_H
+
+#include <grpc/impl/codegen/port_platform.h>
+
+#include <algorithm>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+#include <grpc/slice.h>
+
+// forward-declaring an internal struct, not used publicly.
+struct grpc_slice_buffer;
+
+namespace grpc_event_engine {
+namespace experimental {
+
+// TODO(nnoble): needs implementation
+class SliceBuffer {
+ public:
+  SliceBuffer() { abort(); }
+  explicit SliceBuffer(grpc_slice_buffer*) { abort(); }
+
+  grpc_slice_buffer* RawSliceBuffer() { return slice_buffer_; }
+
+ private:
+  grpc_slice_buffer* slice_buffer_;
+};
+
+// Reservation request - how much memory do we want to allocate?
+class MemoryRequest {
+ public:
+  // Request a fixed amount of memory.
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  MemoryRequest(size_t n) : min_(n), max_(n) {}
+  // Request a range of memory.
+  MemoryRequest(size_t min, size_t max) : min_(std::min(min, max)), max_(max) {}
+
+  // Increase the size by amount
+  MemoryRequest Increase(size_t amount) const {
+    return MemoryRequest(min_ + amount, max_ + amount);
+  }
+
+  size_t min() const { return min_; }
+  size_t max() const { return max_; }
+
+ private:
+  size_t min_;
+  size_t max_;
+};
+
+class BasicMemoryAllocator
+    : public std::enable_shared_from_this<BasicMemoryAllocator> {
+ public:
+  BasicMemoryAllocator() {}
+  virtual ~BasicMemoryAllocator() {}
+
+  BasicMemoryAllocator(const BasicMemoryAllocator&) = delete;
+  BasicMemoryAllocator& operator=(const BasicMemoryAllocator&) = delete;
+
+  // Reserve bytes from the quota.
+  // If we enter overcommit, reclamation will begin concurrently.
+  // Returns the number of bytes reserved.
+  virtual size_t Reserve(MemoryRequest request) = 0;
+
+  // Release some bytes that were previously reserved.
+  virtual void Release(size_t n) = 0;
+
+  // Shutdown this allocator.
+  // Further usage of Reserve() is undefined behavior.
+  virtual void Shutdown() = 0;
+};
+
+class MemoryAllocator {
+ public:
+  // Construct a MemoryAllocator given a BasicMemoryAllocator implementation.
+  // The constructed MemoryAllocator will call BasicMemoryAllocator::Shutdown()
+  // upon destruction.
+  explicit MemoryAllocator(std::shared_ptr<BasicMemoryAllocator> allocator)
+      : allocator_(allocator) {}
+  ~MemoryAllocator() {
+    if (allocator_ != nullptr) allocator_->Shutdown();
+  }
+
+  MemoryAllocator(const MemoryAllocator&) = delete;
+  MemoryAllocator& operator=(const MemoryAllocator&) = delete;
+
+  MemoryAllocator(MemoryAllocator&&) = default;
+  MemoryAllocator& operator=(MemoryAllocator&&) = default;
+
+  // Reserve bytes from the quota.
+  // If we enter overcommit, reclamation will begin concurrently.
+  // Returns the number of bytes reserved.
+  size_t Reserve(MemoryRequest request) { return allocator_->Reserve(request); }
```

Comment:
Just a note, that if `BasicMemoryAllocator::Shutdown` goes away, and that behavior is replaced either with shared_ptr or unique_ptr deletion semantics, then these two methods make `MemoryAllocator` effectively a `BasicMemoryAllocator` implementation as well.

Reply:
Yeah this is not lost on me... The original implementation had just one type in this place, by that in turn relied on OrphanablePtr and InternallyRefCounted, neither of which are suitable for public API.